File: deps/Behavior/test/btunittest/Agent/AgentNodeTest.cpp
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Tencent is pleased to support the open source community by making behaviac available.
//
// Copyright (C) 2015 THL A29 Limited, a Tencent company. All rights reserved.
//
// Licensed under the BSD 3-Clause License (the "License"); you may not use this file except in compliance with
// the License. You may obtain a copy of the License at http://opensource.org/licenses/BSD-3-Clause
//
// Unless required by applicable law or agreed to in writing, software distributed under the License is
// distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and limitations under the License.
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////

#include "AgentNodeTest.h"


BEGIN_ENUM_DESCRIPTION(EnumTest, EnumTest)
{
	DEFINE_ENUM_VALUE(EnumTest_One, "EnumTest_One");
	DEFINE_ENUM_VALUE(EnumTest_OneAfterOne, "EnumTest_OneAfterOne");
}
END_ENUM_DESCRIPTION()

AgentNodeTest::AgentNodeTest()
{
}

AgentNodeTest::~AgentNodeTest()
{
}

void AgentNodeTest::resetProperties()
{
    testVar_0 = -1;
    testVar_1 = -1;
    testVar_2 = -1.0f;
    testVar_3 = -1.0f;
    event_test_var_int = -1;
    event_test_var_bool = false;
    event_test_var_float = -1.0f;
	event_test_var_agent = NULL;
    waiting_timeout_interval = 0;

    action_0_enter_count = 0;
    action_0_exit_count = 0;
    action_1_enter_count = 0;
    action_1_exit_count = 0;
    action_2_enter_count = 0;
    action_2_exit_count = 0;

    testVar_str_0 = "";
	testColor = EnumTest_One;

	m_bCanSee = false;
	m_bTargetValid = false;

	TestFloat2.x = 2.0f;
	TestFloat2.y = 2.0f;
}

namespace UnityEngine
{
    BEGIN_PROPERTIES_DESCRIPTION(GameObject)
    {
        REGISTER_PROPERTY(name);
    }
    END_PROPERTIES_DESCRIPTION()
}

BEGIN_PROPERTIES_DESCRIPTION(AgentNodeTest)
{
    //CLASS_DISPLAYNAME(L"²âÊÔbehaviac::Agent")
    //CLASS_DESC(L"²âÊÔbehaviac::AgentµÄËµÃ÷")
    REGISTER_PROPERTY(testVar_0);
    REGISTER_PROPERTY(testVar_1).DISPLAYNAME(L"testVar_1").DESC(L"testVar_1 property").RANGE(100);
    REGISTER_PROPERTY(testVar_2);
    REGISTER_PROPERTY(testVar_3);
    REGISTER_PROPERTY(waiting_timeout_interval);
    REGISTER_PROPERTY(testVar_str_0);
	REGISTER_PROPERTY(testColor);

    REGISTER_METHOD(setEventVarInt);
    REGISTER_METHOD(setEventVarBool);
    REGISTER_METHOD(setEventVarFloat);
    REGISTER_METHOD(setEventVarAgent);
	REGISTER_METHOD(getConstOne);
	REGISTER_METHOD(getConstThousand);
    REGISTER_METHOD(setTestVar_0);
    REGISTER_METHOD(setTestVar_1);
    REGISTER_METHOD(setTestVar_2);
    REGISTER_METHOD(setTestVar_0_2);
    REGISTER_METHOD(setTestVar_R);
    REGISTER_METHOD(setTestVar_3);
    REGISTER_METHOD(enter_action_0);
    REGISTER_METHOD(exit_action_0);
    REGISTER_METHOD(enter_action_1);
    REGISTER_METHOD(exit_action_1);
    REGISTER_METHOD(enter_action_2);
    REGISTER_METHOD(exit_action_2);
    REGISTER_METHOD(createGameObject);
    REGISTER_METHOD(testGameObject);
	REGISTER_METHOD(createExtendedNode);
	REGISTER_METHOD(testExtendedRefType);
	REGISTER_METHOD(testExtendedStruct);
	REGISTER_METHOD(getExtendedStruct);
	REGISTER_METHOD(getConstExtendedStruct);
    REGISTER_METHOD(switchRef);

	REGISTER_METHOD(Stop);
	REGISTER_METHOD(SelectTarget);

	REGISTER_METHOD(IsTargetValid);
	REGISTER_METHOD(CanSeeEnemy);
	REGISTER_METHOD(Move);
	REGISTER_METHOD(MoveToTarget);
	REGISTER_METHOD(GetRefTree);
	REGISTER_METHOD(initChildAgent);
	REGISTER_METHOD(initChildAgentTest);
}
END_PROPERTIES_DESCRIPTION()

void AgentNodeTest::initChildAgentTest()
{
	ChildNodeTest* testChildAgent = this->getChildAgent<ChildNodeTest>(1, "par_child_agent_1");
	this->SetVariable("par_child_agent_1", testChildAgent);
}

void AgentNodeTest::initChildAgent()
{
    ChildNodeTest* test = this->GetVariable<ChildNodeTest*>("par_child_agent_1");

    test->resetProperties();
    test->testVar_1 = 888;
}

ChildNodeTest::ChildNodeTest(int var_0)
{
	testVar_0 = var_0;
}

ChildNodeTest::~ChildNodeTest()
{
}

float ChildNodeTest::GetConstFloatValue()
{
	return 1000.0f;
}

double ChildNodeTest::GetConstDoubleValue()
{
	return 1000.0;
}

BEGIN_PROPERTIES_DESCRIPTION(ChildNodeTest)
{
	REGISTER_METHOD(GetConstFloatValue);
	REGISTER_METHOD(GetConstDoubleValue);
}
END_PROPERTIES_DESCRIPTION()

File: deps/Behavior/integration/unity/Assets/Scripts/behaviac/BehaviacUnitTest/AgentNodeTest.cs
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Tencent is pleased to support the open source community by making behaviac available.
//
// Copyright (C) 2015 THL A29 Limited, a Tencent company. All rights reserved.
//
// Licensed under the BSD 3-Clause License (the "License"); you may not use this file except in compliance with
// the License. You may obtain a copy of the License at http://opensource.org/licenses/BSD-3-Clause
//
// Unless required by applicable law or agreed to in writing, software distributed under the License is
// distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and limitations under the License.
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////

using UnityEngine;

namespace TestNS
{
    public class Node
	{
        public string name;
	};

    public struct Float2
    {
        public float x;
        public float y;
    };
}

public enum EnumTest
{
    EnumTest_One = 0,
    EnumTest_OneAfterOne = 1,
};


[behaviac.TypeMetaInfo()]
public class AgentNodeTest : behaviac.Agent
{
    [behaviac.MemberMetaInfo()]
    public int testVar_0 = -1;

    [behaviac.MemberMetaInfo("testVar_1", "testVar_1 property", 100)]
    public int testVar_1 = -1;

    [behaviac.MemberMetaInfo()]
    public float testVar_2 = -1.0f;

    [behaviac.MemberMetaInfo()]
    public float testVar_3 = -1.0f;

    [behaviac.MemberMetaInfo()]
    public int waiting_timeout_interval = 0;

    [behaviac.MemberMetaInfo()]
    public string testVar_str_0 = string.Empty;

    [behaviac.MemberMetaInfo()]
    public string testVar_str_1 = string.Empty;

    [behaviac.MemberMetaInfo()]
    public EnumTest testColor = EnumTest.EnumTest_One;

    public bool m_bCanSee = false;

    public bool m_bTargetValid;

    public TestNS.Float2 TestFloat2;

    public int event_test_var_int = -1;
    public bool event_test_var_bool = false;
    public float event_test_var_float = -1.0f;
    public AgentNodeTest event_test_var_agent = null;

    public GameObject testAgentGameObject = null;

    public void resetProperties() {
        testVar_0 = -1;
        testVar_1 = -1;
        testVar_2 = -1.0f;
        testVar_3 = -1.0f;
        event_test_var_int = -1;
        event_test_var_bool = false;
        event_test_var_float = -1.0f;
        event_test_var_agent = null;
        waiting_timeout_interval = 0;

        action_0_enter_count = 0;
        action_0_exit_count = 0;
        action_1_enter_count = 0;
        action_1_exit_count = 0;
        action_2_enter_count = 0;
        action_2_exit_count = 0;

        m_bCanSee = false;
        m_bTargetValid = false;

        testColor = EnumTest.EnumTest_One;
        TestFloat2.x = 2.0f;
        TestFloat2.y = 2.0f;

        testVar_str_0 = string.Empty;
        this.Variables.Clear(true);
    }

    [behaviac.MethodMetaInfo()]
    public void initChildAgentTest()
    {
        var testChildAgent = getChildAgent<ChildNodeTest>("par_child_agent_1");
        this.SetVariable<ChildNodeTest>("par_child_agent_1", testChildAgent);
    }

    public T getChildAgent<T>(string strChildAgentName)
where T : behaviac.Agent
    {
        var childAgent = gameObject.AddComponent<T>();
        return childAgent;
    }

    public void init() {
        base.Init();
        resetProperties();
    }

    public void finl() {
    }

    [behaviac.MethodMetaInfo()]
    public void setEventVarInt(int var) {
        event_test_var_int = var;
    }

    [behaviac.MethodMetaInfo()]
    public void setEventVarBool(bool var) {
        event_test_var_bool = var;
    }

    [behaviac.MethodMetaInfo()]
    public void setEventVarFloat(float var) {
        event_test_var_float = var;
    }

    [behaviac.MethodMetaInfo()]
    public void setEventVarAgent(AgentNodeTest agent)
    {
        event_test_var_agent = agent;
    }

    [behaviac.MethodMetaInfo()]
    public void initChildAgent()
    {
        var test = this.GetVariable<ChildNodeTest>("par_child_agent_1");

        test.resetProperties();
        test.testVar_1 = 888;
    }

    [behaviac.MethodMetaInfo()]
    public int getConstOne() {
        return 1;
    }

    [behaviac.MethodMetaInfo()]
    int getConstThousand(int a, int b)
    {
        return a + b;
    }

    [behaviac.MethodMetaInfo()]
    public void setTestVar_0(int var) {
        testVar_0 = var;
    }

    [behaviac.MethodMetaInfo()]
    public void setTestVar_1(int var) {
        testVar_1 = var;
    }

    [behaviac.MethodMetaInfo()]
    public void setTestVar_2(float var) {
        testVar_2 = var;
    }

    [behaviac.MethodMetaInfo()]
    private void setTestVar_0_2(int var0, float var2) {
        testVar_0 = var0;
        testVar_2 = var2;
    }

    [behaviac.MethodMetaInfo()]
    private float setTestVar_R() {
        return (float)testVar_0 + testVar_2;
    }

    [behaviac.MethodMetaInfo()]
    public void setTestVar_3(float var) {
        testVar_3 = var;
    }

    [behaviac.MethodMetaInfo()]
    public GameObject createGameObject() {
        GameObject go = new GameObject();
        go.name = "HC";
        return go;
    }

    [behaviac.MethodMetaInfo()]
    public void testGameObject(GameObject go) {
        if (go != null)
            testVar_str_0 = go.name;
        else
            testVar_str_0 = "null";
    }

    [behaviac.MethodMetaInfo()]
    TestNS.Node createExtendedNode()
	{
		TestNS.Node n = new TestNS.Node();
		n.name = "NODE";
		return n;
	}

    [behaviac.MethodMetaInfo()]
   	void testExtendedRefType(TestNS.Node go)
	{
        testVar_str_1 = go.name;
	}

    [behaviac.MethodMetaInfo()]
	void testExtendedStruct(ref TestNS.Float2 f)
	{
		f.x = 1.0f;
		f.y = 1.0f;
	}

    [behaviac.MethodMetaInfo()]
    TestNS.Float2 getExtendedStruct()
    {
        return this.TestFloat2;
    }

    [behaviac.MethodMetaInfo()]
    TestNS.Float2 getConstExtendedStruct()
    {
        return this.TestFloat2;
    }

    [behaviac.MethodMetaInfo()]
    public behaviac.EBTStatus switchRef(string refTree) {
        this.btreferencetree(refTree);

        return behaviac.EBTStatus.BT_RUNNING;
    }

    [behaviac.MethodMetaInfo()]
    bool CanSeeEnemy()
	{
		return m_bCanSee;
	}

    [behaviac.MethodMetaInfo()]
    void Stop()
    {
    }

    [behaviac.MethodMetaInfo()]
    bool IsTargetValid()
    {
        return m_bTargetValid;
    }

    [behaviac.MethodMetaInfo()]
    void SelectTarget()
    {
        m_bTargetValid = true;
    }

    [behaviac.MethodMetaInfo()]
	behaviac.EBTStatus Move()
	{
		return behaviac.EBTStatus.BT_RUNNING;
	}

    [behaviac.MethodMetaInfo()]
	behaviac.EBTStatus MoveToTarget()
	{
        return behaviac.EBTStatus.BT_RUNNING;
	}

    // enter action and exit action
    public int action_0_enter_count = 0;

    public int action_0_exit_count = 0;
    public int action_1_enter_count = 0;
    public int action_1_exit_count = 0;
    public int action_2_enter_count = 0;
    public int action_2_exit_count = 0;

    [behaviac.MethodMetaInfo()]
    public bool enter_action_0() {
        action_0_enter_count++;
        return true;
    }

    [behaviac.MethodMetaInfo()]
    public void exit_action_0() {
        action_0_exit_count++;
    }

    [behaviac.MethodMetaInfo()]
    public bool enter_action_1(float f) {
        action_1_enter_count++;
        return true;
    }

    [behaviac.MethodMetaInfo()]
    public void exit_action_1(float f) {
        action_1_exit_count++;
    }

    [behaviac.MethodMetaInfo()]
    public bool enter_action_2(int i, string str) {
        testVar_1 = i;
        testVar_str_0 = str;
        action_2_enter_count++;
        return true;
    }

    [behaviac.MethodMetaInfo()]
    public void exit_action_2(int i, string str) {
        testVar_1 = i;
        testVar_str_0 = str;
        action_2_exit_count++;
    }

    [behaviac.MethodMetaInfo()]
   	public string GetRefTree()  {
		return "node_test/reference_sub_0";
	}

}


[behaviac.TypeMetaInfo()]
public class ChildNodeTest : AgentNodeTest
{
    [behaviac.MethodMetaInfo()]
    public float GetConstFloatValue()
    {
        return 1000.0f;
    }

    [behaviac.MethodMetaInfo()]
    public double GetConstDoubleValue()
    {
        return 1000.0;
    }
}
